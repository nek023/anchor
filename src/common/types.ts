export const ItemType = {
  Bookmark: 'bookmark',
  History: 'history',
  Tab: 'tab',
} as const

export interface BookmarkItem {
  type: typeof ItemType.Bookmark
  title?: string
  url?: string
}

export interface HistoryItem {
  type: typeof ItemType.History
  title?: string
  url?: string
}

export interface TabItem {
  type: typeof ItemType.Tab
  favIconUrl?: string
  title?: string
  url?: string
  tabIndex: number
  windowId: number
}

export type Item = BookmarkItem | HistoryItem | TabItem
