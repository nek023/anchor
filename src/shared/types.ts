export const ItemType = {
  Bookmark: "bookmark",
  History: "history",
  Tab: "tab",
} as const;

export interface BookmarkItem {
  id: string;
  type: typeof ItemType.Bookmark;
  title?: string;
  url?: string;
}

export interface HistoryItem {
  id: string;
  type: typeof ItemType.History;
  title?: string;
  url?: string;
}

export interface TabItem {
  id: string;
  type: typeof ItemType.Tab;
  faviconUrl?: string;
  title?: string;
  url?: string;
  tabIndex: number;
  windowId: number;
}

export type Item = BookmarkItem | HistoryItem | TabItem;
