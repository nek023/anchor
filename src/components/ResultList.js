import React, { PureComponent, PropTypes } from 'react';
import { findDOMNode } from 'react-dom';
import ResultRow from './ResultRow';

export default class ResultList extends PureComponent {
  componentDidUpdate() {
    this.scrollComponentToVisible(this.refs.selected);
  }

  scrollComponentToVisible(component) {
    if (!component) return;
    const node = findDOMNode(component);
    node.scrollIntoViewIfNeeded(false);
  }

  render() {
    const { results, selectedRowIndex } = this.props;

    const resultRows = results.map((result, index) => {
      const selected = (index == selectedRowIndex);
      const ref = selected ? 'selected' : null;

      return <ResultRow key={index} ref={ref} result={result} selected={selected} />;
    });

    return (
      <ul className='item-list'>
        {resultRows}
      </ul>
    );
  }
}

ResultList.propTypes = {
  results:          PropTypes.array,
  selectedRowIndex: PropTypes.number
};

ResultList.defaultProps = {
  results: [],
  selectedRowIndex: 0
};
