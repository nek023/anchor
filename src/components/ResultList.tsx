import React from 'react'
import styled from 'styled-components'
import { Item } from '../types'
import ResultListItem from './ResultListItem'

const ResultListContainer = styled.ul`
  width: 100%;
  height: 432px; /* (46 + 2) * 9 */
  padding: 0;
  margin: 8px 0 0;
  display: block;
  list-style: none;
  overflow-x: hidden;
  overflow-y: auto;
  white-space: nowrap;
  cursor: default;
`

type OptionalProps = {
  onClickItem?: (index: number) => void
  selectedItemIndex?: number
}

type Props = OptionalProps & {
  items: Item[]
}

export default class ResultList extends React.PureComponent<Props> {
  render() {
    const listItems = this.props.items.map((item, index) => {
      const selected = index === this.props.selectedItemIndex

      return (
        <ResultListItem
          index={index}
          item={item}
          key={index}
          onClick={this.handleClick}
          selected={selected}
        />
      )
    })

    return <ResultListContainer>{listItems}</ResultListContainer>
  }

  private handleClick = (index: number) => {
    if (this.props.onClickItem) {
      this.props.onClickItem(index)
    }
  }
}
