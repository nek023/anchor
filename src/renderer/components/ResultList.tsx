import React, { useCallback } from 'react'
import styled from 'styled-components'
import { Item } from '../../common/types'
import { ResultListItem } from './ResultListItem'

const Container = styled.ul`
  width: 100%;
  height: 432px; /* (46 + 2) * 9 */
  padding: 0;
  margin: 8px 0 0;
  display: block;
  list-style: none;
  overflow-x: hidden;
  overflow-y: auto;
  white-space: nowrap;
  cursor: default;
`

interface ResultListProps {
  items: Item[]
  onItemClick: (index: number) => void
  selectedItemIndex: number
}

export const ResultList: React.FC<ResultListProps> = ({
  items,
  onItemClick,
  selectedItemIndex,
}) => {
  const handleClick = useCallback((index: number) => onItemClick(index), [
    onItemClick,
  ])

  return (
    <Container>
      {items.map((item, index) => (
        <ResultListItem
          index={index}
          item={item}
          key={index}
          onClick={handleClick}
          selected={index === selectedItemIndex}
        />
      ))}
    </Container>
  )
}
