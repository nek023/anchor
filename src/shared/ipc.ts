export const MessageType = {
  searchItems: "SEARCH_ITEMS",
  setQuery: "SET_QUERY",
} as const;

export const searchItems = (query: string) => ({
  type: MessageType.searchItems,
  payload: { query },
});

export const setQuery = (query: string) => ({
  type: MessageType.setQuery,
  payload: { query },
});

export type Message = ReturnType<typeof searchItems | typeof setQuery>;

export const sendMessage = (message: Message) =>
  new Promise((resolve, reject) => {
    chrome.runtime.sendMessage(message, (response) => {
      if (response != null) {
        resolve(response);
      } else {
        reject(chrome.runtime.lastError);
      }
    });
  });
