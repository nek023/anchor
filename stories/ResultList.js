import React from 'react';
import { storiesOf } from '@storybook/react';
import { action } from '@storybook/addon-actions';
import { ResultList } from '../src/components';
import { ItemTypes } from '../src/constants';
import { createItem, createResult } from '../test/helpers';
import favicon from './static/favicon.png';

const results = [...Array(10).keys()].map((index) => {
  const item = createItem({
    type: ItemTypes.TAB,
    favIconUrl: favicon
  });
  return createResult({ index, item });
});

const actions = {
  onClickItem: action('onClickItem'),
  onMouseEnterItem: action('onMouseEnterItem')
};

storiesOf('ResultList', module)
  .add('with results', () => (
    <ResultList results={results} {...actions} />
  ))
  .add('with no results', () => (
    <ResultList {...actions} />
  ));
