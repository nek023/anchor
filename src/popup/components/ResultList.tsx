import React, { useCallback } from "react";
import { Item } from "../../shared/types";
import { ResultListItem } from "./ResultListItem";

interface ResultListProps {
  items: Item[];
  selectedItemIndex: number;
  onItemClick?: (index: number) => void;
}

export const ResultList: React.FC<ResultListProps> = ({
  items,
  onItemClick,
  selectedItemIndex,
}) => {
  const handleClick = useCallback(
    (item: Item) => {
      if (onItemClick == null) return;
      const index = items.findIndex((i) => i.id === item.id);
      onItemClick(index);
    },
    [items, onItemClick],
  );

  return (
    <div className="w-full flex flex-col grow overflow-y-auto">
      {items.map((item, index) => (
        <ResultListItem
          key={index}
          item={item}
          selected={index === selectedItemIndex}
          onClick={handleClick}
        />
      ))}
    </div>
  );
};
