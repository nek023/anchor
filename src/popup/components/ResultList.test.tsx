import React from "react";
import { fireEvent, render } from "@testing-library/react";
import { BookmarkItem, ItemType } from "../../shared/types";
import { ResultList } from "./ResultList";

describe("ResultList", () => {
  const items: BookmarkItem[] = [
    {
      id: "bookmark-1",
      type: ItemType.Bookmark,
      title: "item1",
      url: "https://example.com",
    },
    {
      id: "bookmark-2",
      type: ItemType.Bookmark,
      title: "item2",
      url: "https://example.com",
    },
  ];

  beforeEach(() => {
    global.chrome = {
      runtime: {
        getURL: (path: string) => {
          return `chrome-extension://EXTENSION_ID${path}`;
        },
      },
      // eslint-disable-next-line @typescript-eslint/no-explicit-any
    } as any;
  });

  test("matches snapshot", () => {
    const { asFragment } = render(
      <ResultList items={items} selectedItemIndex={0} />
    );

    expect(asFragment()).toMatchSnapshot();
  });

  test("calls onClickItem when the item has been clicked", () => {
    const handleItemClick = jest.fn();
    const { getByText } = render(
      <ResultList
        items={items}
        selectedItemIndex={0}
        onItemClick={handleItemClick}
      />
    );

    fireEvent.click(getByText("item2"));
    expect(handleItemClick).toBeCalledWith(1);
  });
});
