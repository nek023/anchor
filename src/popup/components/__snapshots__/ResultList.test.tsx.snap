// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResultList matches snapshot 1`] = `
<DocumentFragment>
  .c2 {
  width: 30px;
  min-width: 30px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  -webkit-box-pack: center;
  -webkit-justify-content: center;
  -ms-flex-pack: center;
  justify-content: center;
}

.c4 {
  margin-left: 8px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  overflow: hidden;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c3 {
  width: 24px;
  height: 24px;
}

.c6 {
  color: #333333;
  font-size: 15px;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c8 {
  color: #9999aa;
  margin-top: 2px;
  font-size: 13px;
  font-weight: lighter;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
}

.c1 {
  width: 100%;
  height: 46px;
  padding: 3px;
  margin-bottom: 2px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  background-color: #0067ff;
}

.c1 .c5 {
  color: #fafafa;
}

.c1 .c7 {
  color: #bcd5fb;
}

.c9 {
  width: 100%;
  height: 46px;
  padding: 3px;
  margin-bottom: 2px;
  overflow: hidden;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
}

.c0 {
  width: 100%;
  height: 432px;
  padding: 0;
  margin: 8px 0 0;
  display: block;
  list-style: none;
  overflow-x: hidden;
  overflow-y: auto;
  white-space: nowrap;
  cursor: default;
}

<ul
    class="c0"
  >
    <li
      class="c1"
    >
      <div
        class="c2"
      >
        <img
          class="c3"
          src="chrome-extension://EXTENSION_ID/_favicon/?pageUrl=https%3A%2F%2Fexample.com&size=32"
        />
      </div>
      <div
        class="c4"
      >
        <div
          class="c5 c6"
        >
          item1
        </div>
        <div
          class="c7 c8"
        >
          https://example.com
        </div>
      </div>
    </li>
    <li
      class="c9"
    >
      <div
        class="c2"
      >
        <img
          class="c3"
          src="chrome-extension://EXTENSION_ID/_favicon/?pageUrl=https%3A%2F%2Fexample.com&size=32"
        />
      </div>
      <div
        class="c4"
      >
        <div
          class="c5 c6"
        >
          item2
        </div>
        <div
          class="c7 c8"
        >
          https://example.com
        </div>
      </div>
    </li>
  </ul>
</DocumentFragment>
`;
